File: StartUpExample/StartUpExampleJava/app/src/main/java/com/tinlone/startupexamplejava/dependencies/Cache.java
package com.tinlone.startupexamplejava.dependencies;

import com.tinlone.startupexamplejava.dependencies.database.DatabaseProxy;

public class Cache {

    static DatabaseProxy databaseProxy;

    private Cache() {
        Logger.log("loglog", "Cache initialized");
    }

    public static void initialize(DatabaseProxy db) {
        databaseProxy = db;
    }

    public static Cache getInstance() {
        return Holder.INSTANCE;
    }

    private static final class Holder {
        public static final Cache INSTANCE = new Cache();
    }

}

File: StartUpExample/StartUpExampleKotlin/app/src/main/java/com/tinlone/startupexamplekotlin/dependencies/Cache.kt
package com.tinlone.startupexamplekotlin.dependencies

import com.tinlone.startupexamplekotlin.dependencies.database.DatabaseProxy

class Cache private constructor() {
    private object Holder {
        val instance = Cache()
    }

    companion object {
        var databaseProxy: DatabaseProxy? = null
        fun initialize(db: DatabaseProxy?) {
            databaseProxy = db
        }

        fun getInstance() :Cache{
            return Holder.instance
        }
    }

    init {
        Logger.log("loglog", "Cache initialized")
    }
}